pkg/controllers/webhookcerts/cert_requester.go

```diff
@@ -0,0 +1,380 @@
+		key, err = pemDecodeRSAPrivate(keyPEM)
+		if err != nil {
+			klog.Error("failed to decode the private key used for the openshift-authenticator, new key will be generated: %v", err)
+			// delete the secret and regenerate it in the next step with a new key
+			err = c.secretClient.Delete(ctx, certSecretName, metav1.DeleteOptions{})
+			if err != nil {
+				return nil, nil, false, err
+			}
+		}
+	}
+	if key == nil {
+		key, err = rsa.GenerateKey(rand.Reader, 4096)
+		if err != nil {
+			return nil, nil, false, fmt.Errorf("failed to generate a private key: %w", err)
+		}
+		keyPEM, err := pemEncodeRSAPrivate(key)
+		if err != nil {
+			return nil, nil, false, err
+		}
+
+		if err := c.secretClient.Delete(ctx, certSecretName, metav1.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+			return nil, nil, false, err
+		}
+
+		_, err = c.secretClient.Create(ctx, &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: certSecretName,
+			},
+			Data: map[string][]byte{
+				"tls.key": keyPEM,
+			},
+		}, metav1.CreateOptions{})
+
+		return key, nil, true, err
+	}
+
+	certPEM := certSecret.Data["tls.crt"]
+	if len(certPEM) > 0 {
+		cert, err = pemDecodeCert(certPEM)
+
+		// delete tls.crt if it's malformed or its pub key does not match the private key
+		var deleteCert bool
+		if err != nil {
+			klog.Error("failed to decode the certificate used for the openshift-authenticator, new certificate will be requested: %v", err)
+			deleteCert = true
+		} else if !key.PublicKey.Equal(cert.PublicKey.(*rsa.PublicKey)) {
+			klog.Error("the public key of the certificate used for the openshift-authenticator does not match the stored private key, new certificate will be requested: %v", err)
+			deleteCert = true
+		}
+
+		if deleteCert {
+			certSecretCopy := certSecret.DeepCopy()
+			delete(certSecretCopy.Data, "tls.crt")
+			_, err := c.secretClient.Update(ctx, certSecretCopy, metav1.UpdateOptions{})
+			return key, nil, false, err
+		}
+	}
+
+	return key, cert, false, nil
+}
+
+// resubmitCSR deletes the old CSR API object and creates a new one for the given key
+func (c *webhookAuthenticatorCertRequester) resubmitCSR(ctx context.Context, key *rsa.PrivateKey) error {
+	err := c.csrClient.Delete(ctx, csrName, metav1.DeleteOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	csrDER, err := x509.CreateCertificateRequest(rand.Reader,
+		&x509.CertificateRequest{
+			Subject: pkix.Name{CommonName: "system:serviceaccount:openshift-oauth-apiserver:openshift-authenticator"},
+		},
+		key,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create a CSR: %w", err)
+	}
+
+	csrPEM, err := pemEncodeCSR(csrDER)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.csrClient.Create(ctx,
+		&certapiv1.CertificateSigningRequest{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: csrName,
+			},
+			Spec: certapiv1.CertificateSigningRequestSpec{
+				Request:    csrPEM,
+				SignerName: certapiv1.KubeAPIServerClientSignerName,
```

Comment:
@stlaz how will `oauth-apiserver` verify a cert signed by `KubeAPIServerClientSignerName`?

Haven't seen it accepting a custom client CA.

Reply:
that's not a custom CA, it's the CA already used for client certificate verification

Reply:
Haven't seen `oauth-apiserver` accepting any CA :)

Reply:
https://github.com/openshift/oauth-apiserver/blob/44c67e17695c437ea96c3715f498728077b60322/vendor/k8s.io/apiserver/pkg/server/options/authentication.go#L305-L313

Reply:
thanks, didn't know it contains `client-ca-file`